src/networking.c

```diff
@@ -1590,8 +1590,99 @@ int _writeToClient(client *c, ssize_t *nwritten) {
         return C_OK;
     }
 
-    if (c->bufpos > 0) {
-        *nwritten = connWrite(c->conn,c->buf+c->sentlen,c->bufpos-c->sentlen);
+    int nbuf = listLength(c->reply);
+    if (nbuf > 0) {
+        if (c->bufpos > 0) ++nbuf; 
+        if (nbuf > 1) {
+            struct iovec iov[nbuf > IOV_MAX ? IOV_MAX : nbuf];
+            int iovcnt = 0;
+            size_t cum_bytes = 0;
+            /* If the successive buffer of responses is not empty, 
+             * add it to the buffer queue of writev() as well. */
+            if (c->bufpos > 0) {
+                iov[iovcnt].iov_base = c->buf + c->sentlen;
+                iov[iovcnt].iov_len = c->bufpos - c->sentlen;
+                cum_bytes += iov[iovcnt++].iov_len;
+            }
+            /* The first node of reply list might be incomplete from the last call,
+             * thus it needs to be calibrated to get the actual data address and length. */
+            size_t offset = c->bufpos > 0 ? 0 : c->sentlen;
+            listIter iter;
+            listNode *next;
+            clientReplyBlock *o;
+            listRewind(c->reply, &iter);
+            while ((next = listNext(&iter)) && iovcnt < IOV_MAX && cum_bytes < NET_MAX_WRITES_PER_EVENT) {
+                o = listNodeValue(next);
+                if (o->used == 0) { /* empty node, just release it and skip. */
+                    c->reply_bytes -= o->size;
+                    listDelNode(c->reply, next);
+                    continue;
+                }
+
+                iov[iovcnt].iov_base = o->buf + offset;
+                iov[iovcnt].iov_len = o->used - offset;
+                cum_bytes += iov[iovcnt++].iov_len;
+                offset = 0;
+            }
+            if (iovcnt == 0) return C_OK;
+            *nwritten = connWritev(c->conn, iov, iovcnt);
+            if (*nwritten <= 0) return C_ERR;
+
+            /* Locate the new node which has leftover data and
+            * release all nodes in front of it. */
+            ssize_t remaining = *nwritten;
+            if (c->bufpos > 0) { /* deal with successive buffer of responses first. */
+                c->sentlen += remaining;
+
+                /* If the buffer was sent, set bufpos to zero to continue with
+                * the remainder of the reply. */
+                if ((int)c->sentlen >= c->bufpos) {
+                    c->bufpos = 0;
+                    c->sentlen = 0;
+                }
+                remaining -= iov[0].iov_len;
```

Comment:
if we didn't write the full buffer, don't we need to break out?
isn't doing `remaining -=` on the entire size of `iov[0].iov_len` too much?

Reply:
> isn't doing remaining -= on the entire size of iov[0].iov_len too much?

Yes, then `remaining` is less than 0, and it will skip the below `while (remaining > 0)`, which is the same effect as breaking out.